core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java

```diff
@@ -207,10 +207,22 @@ public void execute(Runnable command) {
       return continueOnNonBlockingThread(stage.thenRunAsync(runnable, blockingExecutor), traceId);
    }
 
+   @Override
+   public <I, O> CompletionStage<O> thenComposeBlocking(CompletionStage<? extends I> stage,
+         Function<? super I, ? extends CompletionStage<O>> function, Object traceId) {
+      if (CompletionStages.isCompletedSuccessfully(stage) && isCurrentThreadBlocking()) {
+         if (log.isTraceEnabled()) {
+            log.tracef("Invoked thenComposeBlocking on a blocking thread, joining %s in same blocking thread", traceId);
+         }
+         return stage.thenCompose(function);
```

Comment:
you probably could do a `join()` and invoke `function.apply()` directly.

Reply:
Sure, should be slightly less allocations for only marginally more code.